### ISSUE TITLE
Support negative domain slices in queries

### ISSUE THREAD

**Issue by abthompson** [TileDB-Inc/TileDB-Py#102]
TileDB: `1.4.0`
TileDB-Py: `0.3.0`

I'm investigating using TileDB(-Py) to store the following sparse geospatial data:
- 2.5D (from GeoTIFFs): latitude, longitude, elevation
- 3D (from LAS point clouds): latitude, longitude, elevation, intensity

These types of domains inherently depend on negative, real domain values. [TileDB documentation](https://docs.tiledb.io/en/stable/tutorials/neg-real.html) indicates negative and real domain values are supported, which is encouraging.

The support/use of negative domain values also implies one might need the ability to use slices with negative values when performing queries. Unfortunately, it seems queries with negative domain values are still unsupported. Attempting to use negative domain values in slices results in an `IndexError: index out of bounds <todo>`, thrown [here](https://github.com/TileDB-Inc/TileDB-Py/blob/0.3.0/tiledb/libtiledb.pyx#L2873).

I've tried to mitigate the issue by setting the domain dimensions to be non-negative (latitude: 0-180; longitude: 0-360) and using conversion methods to ensure domain values are non-negative when writing, then converting to their real values for queries or return query results. The following would be examples of the aforementioned conversion methods:
```py
def to_domain_latitude(real_latitude):
    '''Convert real latitude to domain latitude'''
    return real_latitude + 180.0

def to_real_latitude(domain_latitude):
    '''Convert domain latitude to real latitude'''
    return domain_latitude - 180.0
```
However, it would be simplest for us developers to store the domain values in their native ranges/formats.

**Comment by stavrospapadopoulos** [comment]
Thanks for reaching out. Negative domain values are supported in the core library and most of our APIs. Unfortunately, our Python API currently performs slicing through operator `[]`, which, similar to numpy indexing, interprets negative indices as counting from the end of the array domain. However, we are planning to add negative indexing support to our Python API very soon. 

As a side note, since you are working with real lat/lon values, please be aware of [this issue](https://github.com/TileDB-Inc/TileDB-Py/issues/81), just in case you encounter it as well. We plan to fix it early next week.

**Comment by thompsonab** [comment]
@stavrospapadopoulos Thank you for your response. I look forward to the negative indexing support. And I apologize for not just commenting on #2, but I wasn't sure if negative indices/ranges in slices would qualify, or if that issue were more for single-value indices.

Since you mentioned it's supported in other APIs, I've successfully used the C++ API to perform similar tests with the "2.5D" GeoTIFF data I mentioned previously. It's just that we're particularly interested in the Python API as we're investigating integrating this as a backend data storage tool for existent Python-based utilities.

Thank you for the heads-up on the issue you mentioned. I've actually noticed some segfaults, but I wasn't convinced of its origin (PDAL, TileDB, or my own code). It's possible I'm running into the same problem, but it's tough to say for sure either way. I'll be happy to try an updated version when it's available.

To be fair, I'm also not certain I'm using the API the correctly, but that's more on me taking the time to read through how to choose the appropriate tiling schemes, filters, etc.

**Comment by jakebolewski** [comment]
Negative domain's were added / fully supported after the binding was developed, so we have to play catch up with the functionality in the core library.   We could support negative indexing easily but it would not match Numpy semantics.   I don't know what is the "least surprising" behavior, not being compatible with Numpy or just mapping negative indices to the TileDB domain subquery.   @abthompson thoughts?  I think this is more of an api design issue, we can easily accept both or default to one behavior.

**Comment by abthompson** [comment]
### TL;DR
As long as there's a way If there's another way to perform the queries that doesn't rely on the assumption that negative domain values are intended to wrap around from the upper boundary, I have no preference on whether negative domain values are supported via slicing/ the `__getitem__` interface.

### VERBOSE=1
My perspective is colored through the lenses of a C++ background with far less experience in Python, so you may not want to put much stock in it...

I completely understand/appreciate that the library as a whole is continually improving and there's inherently some catching up to do across language APIs. Unfortunately, being able to perform queries with negative domain values using the C++ API but not in the Python API is a frustrating/confusing disconnect as a potential user.

This is all I need for my purposes at the moment: to be able to read/write data using negative domain values. I've been able to do so via the C++ API's `Query` class (via `set_coordinates` and `set_buffer`, specifically). From my perspective, it doesn't matter if I have to go through a similar interface on the Python side. I've done it already, so it would be familiar. Slices are convenient/common, but it can quickly get confusing for you developers to determine when to treat negative values as wrap-around or literal values.

I apologize for not explicitly answering your question... I honestly can't (shouldn't?) push you in either direction at the moment.
